### PR TITLE
own: basic auto-indexing for background signal calculations

### PR DESCRIPTION
**Pull request by coury-clark**
Closes https://github.com/sourcegraph/sourcegraph/issues/50939

Adds basic auto-indexing and scheduling for own background jobs. Jobs are partitioned by repo_id. It's all very light and non-committal, so we can change this structure as required.

Jobs are each individually feature flagged with the prefix `own-background-index-repo`, for example `own-background-index-repo-recent-contributors`.

Jobs are fully observable with standard metrics for background workers, the dashboards can be found on the `Worker` dashboard.

The queue that actually runs the jobs is rate limited, which is configurable through the site config.

```
own.background.repoIndexRateLimit
own.background.repoIndexRateBurstLimit
own.background.repoIndexConcurrencyLimit (requires restart)
```

## Test plan
Enable the feature flag and check worker logs, example:

```
[worker] INFO worker.own-repo-indexing-queue.own.background.index.scheduler.recent-contributors background/scheduler.go:84 Scheduling repo indexes for own job {"job-name": "recent-contributors"}
[worker] INFO worker.own-repo-indexing-queue.own.background.index.scheduler.recent-contributors background/scheduler.go:96 Own index job scheduled {"job-name": "recent-contributors", "row-count": 11}
[worker] INFO worker.own-repo-indexing-queue workerutil/worker.go:326 Dequeued record for processing {"TraceId": "2f83d932404f2fc0e48ea42b708cbfb8", "SpanId": "b492d1d7607c5ef5", "name": "own_background_worker", "id": 69}
[worker] INFO worker.own-repo-indexing-queue.Handle background/background.go:182 Hello from the own background processor: {"ID":69,"State":"processing","FailureMessage":null,"QueuedAt":"2023-04-24T21:02:13.564001Z","StartedAt":"2023-04-24T21:02:23.546593Z","FinishedAt":null,"ProcessAfter":null,"NumResets":0,"NumFailures":0,"LastHeartbeatAt":"2023-04-24T21:02:23.546593Z","ExecutionLogs":null,"WorkerHostname":"Courys-MacBook-Pro-3.local","Cancel":false,"RepoId":3,"JobType":1} {"TraceId": "2f83d932404f2fc0e48ea42b708cbfb8", "SpanId": "5df6ef42d43b7094", "handle": {}}
[worker] INFO worker.own-repo-indexing-queue workerutil/worker.go:326 Dequeued record for processing {"name": "own_background_worker", "id": 70}
[worker] INFO worker.own-repo-indexing-queue.Handle background/background.go:182 Hello from the own background processor: {"ID":70,"State":"processing","FailureMessage":null,"QueuedAt":"2023-04-24T21:02:13.564001Z","StartedAt":"2023-04-24T21:02:23.554898Z","FinishedAt":null,"ProcessAfter":null,"NumResets":0,"NumFailures":0,"LastHeartbeatAt":"2023-04-24T21:02:23.554898Z","ExecutionLogs":null,"WorkerHostname":"Courys-MacBook-Pro-3.local","Cancel":false,"RepoId":4,"JobType":1} {"handle": {}}
[worker] INFO worker.own-repo-indexing-queue workerutil/worker.go:326 Dequeued record for processing {"TraceId": "025c8d0e7942f14ca87585a4a280a776", "SpanId": "e957e221d136ed6c", "name": "own_background_worker", "id": 71}
[worker] INFO worker.own-repo-indexing-queue.Handle background/background.go:182 Hello from the own background processor: {"ID":71,"State":"processing","FailureMessage":null,"QueuedAt":"2023-04-24T21:02:13.564001Z","StartedAt":"2023-04-24T21:02:23.556934Z","FinishedAt":null,"ProcessAfter":null,"NumResets":0,"NumFailures":0,"LastHeartbeatAt":"2023-04-24T21:02:23.556934Z","ExecutionLogs":null,"WorkerHostname":"Courys-MacBook-Pro-3.local","Cancel":false,"RepoId":7,"JobType":1} {"TraceId": "025c8d0e7942f14ca87585a4a280a776", "SpanId": "1e96598b0b05f23d", "handle": {}}
[worker] INFO worker.own-repo-indexing-queue workerutil/worker.go:326 Dequeued record for processing {"TraceId": "a80c72cfd675ee09db9fe6f4194ef5f6", "SpanId": "1bbb106bd9912b49", "name": "own_background_worker", "id": 72}
[worker] INFO worker.own-repo-indexing-queue.Handle background/background.go:182 Hello from the own background processor: {"ID":72,"State":"processing","FailureMessage":null,"QueuedAt":"2023-04-24T21:02:13.564001Z","StartedAt":"2023-04-24T21:02:23.559786Z","FinishedAt":null,"ProcessAfter":null,"NumResets":0,"NumFailures":0,"LastHeartbeatAt":"2023-04-24T21:02:23.559786Z","ExecutionLogs":null,"WorkerHostname":"Courys-MacBook-Pro-3.local","Cancel":false,"RepoId":6,"JobType":1} {"TraceId": "a80c72cfd675ee09db9fe6f4194ef5f6", "SpanId": "354c9adfa90f1a6c", "handle": {}}
[worker] INFO worker.own-repo-indexing-queue workerutil/worker.go:326 Dequeued record for processing {"name": "own_background_worker", "id": 73}
[worker] INFO worker.own-repo-indexing-queue.Handle background/background.go:182 Hello from the own background processor: {"ID":73,"State":"processing","FailureMessage":null,"QueuedAt":"2023-04-24T21:02:13.564001Z","StartedAt":"2023-04-24T21:02:23.560648Z","FinishedAt":null,"ProcessAfter":null,"NumResets":0,"NumFailures":0,"LastHeartbeatAt":"2023-04-24T21:02:23.560648Z","ExecutionLogs":null,"WorkerHostname":"Courys-MacBook-Pro-3.local","Cancel":false,"RepoId":5,"JobType":1} {"handle": {}}
[worker] INFO worker.own-repo-indexing-queue workerutil/worker.go:326 Dequeued record for processing {"TraceId": "43ad71d53a938a0baccbe9511623c75e", "SpanId": "e7af1bf0a7b25dfc", "name": "own_background_worker", "id": 74}
[worker] INFO worker.own-repo-indexing-queue workerutil/worker.go:326 Dequeued record for processing {"name": "own_background_worker", "id": 75}
[worker] INFO worker.own-repo-indexing-queue workerutil/worker.go:326 Dequeued record for processing {"TraceId": "df4d2f280ebdadcf516e81d1213d5a5f", "SpanId": "8604bd7f483de3ce", "name": "own_background_worker", "id": 76}
[worker] INFO worker.own-repo-indexing-queue workerutil/worker.go:326 Dequeued record for processing {"name": "own_background_worker", "id": 77}
[worker] INFO worker.own-repo-indexing-queue workerutil/worker.go:326 Dequeued record for processing {"name": "own_background_worker", "id": 78}
[worker] INFO worker.own-repo-indexing-queue.Handle background/background.go:182 Hello from the own background processor: {"ID":74,"State":"processing","FailureMessage":null,"QueuedAt":"2023-04-24T21:02:13.564001Z","StartedAt":"2023-04-24T21:02:23.562546Z","FinishedAt":null,"ProcessAfter":null,"NumResets":0,"NumFailures":0,"LastHeartbeatAt":"2023-04-24T21:02:23.562546Z","ExecutionLogs":null,"WorkerHostname":"Courys-MacBook-Pro-3.local","Cancel":false,"RepoId":1,"JobType":1} {"TraceId": "43ad71d53a938a0baccbe9511623c75e", "SpanId": "2571c40fedc55398", "handle": {}}
[worker] INFO worker.own-repo-indexing-queue workerutil/worker.go:326 Dequeued record for processing {"name": "own_background_worker", "id": 79}
[worker] INFO worker.own-repo-indexing-queue.Handle background/background.go:182 Hello from the own background processor: {"ID":75,"State":"processing","FailureMessage":null,"QueuedAt":"2023-04-24T21:02:13.564001Z","StartedAt":"2023-04-24T21:02:23.563656Z","FinishedAt":null,"ProcessAfter":null,"NumResets":0,"NumFailures":0,"LastHeartbeatAt":"2023-04-24T21:02:23.563656Z","ExecutionLogs":null,"WorkerHostname":"Courys-MacBook-Pro-3.local","Cancel":false,"RepoId":9,"JobType":1} {"handle": {}}
[worker] INFO worker.own-repo-indexing-queue.Handle background/background.go:182 Hello from the own background processor: {"ID":76,"State":"processing","FailureMessage":null,"QueuedAt":"2023-04-24T21:02:13.564001Z","StartedAt":"2023-04-24T21:02:23.564807Z","FinishedAt":null,"ProcessAfter":null,"NumResets":0,"NumFailures":0,"LastHeartbeatAt":"2023-04-24T21:02:23.564807Z","ExecutionLogs":null,"WorkerHostname":"Courys-MacBook-Pro-3.local","Cancel":false,"RepoId":10,"JobType":1} {"TraceId": "df4d2f280ebdadcf516e81d1213d5a5f", "SpanId": "09e634f31d4490b7", "handle": {}}
[worker] INFO worker.own-repo-indexing-queue.Handle background/background.go:182 Hello from the own background processor: {"ID":77,"State":"processing","FailureMessage":null,"QueuedAt":"2023-04-24T21:02:13.564001Z","StartedAt":"2023-04-24T21:02:23.56597Z","FinishedAt":null,"ProcessAfter":null,"NumResets":0,"NumFailures":0,"LastHeartbeatAt":"2023-04-24T21:02:23.56597Z","ExecutionLogs":null,"WorkerHostname":"Courys-MacBook-Pro-3.local","Cancel":false,"RepoId":8,"JobType":1} {"handle": {}}
[worker] INFO worker.own-repo-indexing-queue.Handle background/background.go:182 Hello from the own background processor: {"ID":78,"State":"processing","FailureMessage":null,"QueuedAt":"2023-04-24T21:02:13.564001Z","StartedAt":"2023-04-24T21:02:23.566886Z","FinishedAt":null,"ProcessAfter":null,"NumResets":0,"NumFailures":0,"LastHeartbeatAt":"2023-04-24T21:02:23.566886Z","ExecutionLogs":null,"WorkerHostname":"Courys-MacBook-Pro-3.local","Cancel":false,"RepoId":11,"JobType":1} {"handle": {}}
[worker] INFO worker.own-repo-indexing-queue.Handle background/background.go:182 Hello from the own background processor: {"ID":79,"State":"processing","FailureMessage":null,"QueuedAt":"2023-04-24T21:02:13.564001Z","StartedAt":"2023-04-24T21:02:23.605941Z","FinishedAt":null,"ProcessAfter":null,"NumResets":0,"NumFailures":0,"LastHeartbeatAt":"2023-04-24T21:02:23.605941Z","ExecutionLogs":null,"WorkerHostname":"Courys-MacBook-Pro-3.local","Cancel":false,"RepoId":2,"JobType":1} {"handle": {}}

```

### CODE REVIEW

File: enterprise/internal/own/background/background.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+	})
+
+	task := handler{
+		workerStore: workerStore,
+	}
+
+	worker := dbworker.NewWorker(ctx, workerStore, workerutil.Handler[*Job](&task), workerutil.WorkerOptions{
+		Name:              name,
+		Description:       "Sourcegraph own background processing partitioned by repository",
+		NumHandlers:       1,
```
Comment (coury-clark) [inline enterprise/internal/own/background/background.go:133]: todo: configuration

File: enterprise/internal/own/background/background.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+func makeWorker(ctx context.Context, db database.DB, observationCtx *observation.Context) (*workerutil.Worker[*Job], *dbworker.Resetter[*Job], dbworkerstore.Store[*Job]) {
+	name := "own_background_worker"
+
+	workerStore := dbworkerstore.New(observationCtx, db.Handle(), dbworkerstore.Options[*Job]{
+		Name:              fmt.Sprintf("%s_store", name),
+		TableName:         tableName,
+		ColumnExpressions: jobColumns,
+		Scan:              dbworkerstore.BuildWorkerScan(scanJob),
+		OrderByExpression: sqlf.Sprintf("id"), // processes oldest records first
```
Comment (coury-clark) [inline enterprise/internal/own/background/background.go:119]: for now all jobs will share the same queue. If we need to spawn off separate workers we can turn this table into a view and create discrete handlers for special cases

**Review by cbart — APPROVED** [review]
Love it! Do you think a test that schedules a new job and verifies it's being picked up and processed could be feasible?

**Comment by coury-clark** [comment]
@cbart yep, I have a bit more work to do here and that will be part of it!

**Comment by cbart** [comment]
Oh and let's make sure it's behind a feature-flag. I don't think we should enable that on dotcom :)

**Comment by coury-clark** [comment]
@cbart I've put each indexer behind a feature flag for now - in the future we will probably want to control this through some kind of UI, this works at the moment.

**Pull request merged**